stdlib/public/core/UnsafeBufferPointer.swift.gyb

```diff
@@ -425,12 +460,134 @@ extension UnsafeMutableBufferPointer {
+    where S:Sequence, S.Element == Element {
+    return source._copyContents(assigningTo: self)
+  }
+
+  // Warning: All buffer elements must be initialized before calling this. Note 
+  // that invoking initialize(from:) may not initialize all elements. 
+  // Assigning to part of the buffer must be done using the moveAssign(from:count:) 
+  // method on the buffer’s baseAddress. 
+  @_inlineable
+  public func moveAssign(from source: UnsafeMutableBufferPointer<Element>) {
+    _debugPrecondition(source.count >= count,
+                      "${Self}.moveAssign(from:) source does not have enough elements")
+    guard let dstStart = _position,
+          let srcStart = source._position else {
+      return
+    }
+
+    dstStart.moveAssign(from: srcStart, count: source._end! - srcStart)
+  }
+
+  @_inlineable
+  public func moveInitialize(from source: UnsafeMutableBufferPointer<Element>) {
+    _debugPrecondition(source.count >= count,
+                      "${Self}.moveInitialize(from:) source does not have enough elements")
+    guard let dstStart = _position,
+          let srcStart = source._position else {
+      return
+    }
+
+    dstStart.moveInitialize(from: srcStart, count: source._end! - srcStart)
+  }
+
+  // Warning: All buffer elements must be initialized before calling this. Note 
+  // that invoking initialize(from:) may not initialize all elements. 
+  // Deinitializing part of the buffer must be done using the deinitialize(count:) 
+  // method on the buffer’s baseAddress. 
+  @_inlineable
+  public func deinitialize() {
+    guard let dstStart = _position else {
+      return
+    }
+
+    dstStart.deinitialize(count: _end! - dstStart)
+  }
+
+% end
+
+  @_inlineable
+  public
+  func withMemoryRebound<T, Result>(
+    to type: T.Type, _ body: (${Self}<T>) throws -> Result) rethrows -> Result {
+    if let base = _position {
+      let capacity = count * MemoryLayout<Element>.stride / MemoryLayout<T>.stride
+      Builtin.bindMemory(base._rawValue, capacity._builtinWordValue, T.self)
+      defer {
+        Builtin.bindMemory(base._rawValue, capacity._builtinWordValue, Element.self)
+      }
+
+      return try body(${Self}<T>( start: Unsafe${Mutable}Pointer<T>(base._rawValue),
+                                  count: capacity))
+    }
+    else {
+      return try body(${Self}<T>(start: nil, count: 0))
+    }
+  }
+
+  /// A pointer to the first element of the buffer.
+  ///
+  /// If the `baseAddress` of this buffer is `nil`, the count is zero. However,
+  /// a buffer can have a `count` of zero even with a non-`nil` base address.
+  @_inlineable
+  public var baseAddress: Unsafe${Mutable}Pointer<Element>? {
+    return _position
+  }
+
+  /// The number of elements in the buffer.
+  ///
+  /// If the `baseAddress` of this buffer is `nil`, the count is zero. However,
+  /// a buffer can have a `count` of zero even with a non-`nil` base address.
+  @_inlineable
+  public var count: Int {
+    switch(_position, _end) {
```

Comment:
Style nits: space after `switch`; `default` instead of `case _` (which isn't used in stdlib unless i'm mistaken).

More generally, though, isn't this more idiomatically just:

```swift
guard let position = _position, let end = _end else { return 0 }
return end - position
```